pkg/sa/agent.go

```diff
@@ -0,0 +1,60 @@
+package sa
+
+import (
+	"context"
+
+	"github.com/rancher/system-agent/pkg/applyinator"
+	"github.com/rancher/system-agent/pkg/config"
+	"github.com/rancher/system-agent/pkg/image"
+	"github.com/rancher/system-agent/pkg/k8splan"
+	"github.com/rancher/system-agent/pkg/localplan"
+	"github.com/rancher/system-agent/pkg/version"
+	"github.com/sirupsen/logrus"
+)
+
+type Agent struct {
+	cfg *config.AgentConfig
+}
+
+func (a *Agent) Run(ctx context.Context) error {
+
+	if a.cfg == nil {
+		return nil
+	}
+
+	logrus.Infof("Rancher System Agent version %s is starting", version.FriendlyVersion())
+
+	if !a.cfg.LocalEnabled && !a.cfg.RemoteEnabled {
+		logrus.Fatalf("Local and remote were both not enabled. Exiting, as one must be enabled.")
+	}
+
+	logrus.Infof("Using directory %s for work", a.cfg.WorkDir)
+
+	imageUtil := image.NewUtility(a.cfg.ImagesDir, a.cfg.ImageCredentialProviderConfig, a.cfg.ImageCredentialProviderBinDir, a.cfg.AgentRegistriesFile)
+	applier := applyinator.NewApplyinator(a.cfg.WorkDir, a.cfg.PreserveWorkDir, a.cfg.AppliedPlanDir, imageUtil)
+
+	if a.cfg.RemoteEnabled {
+		logrus.Infof("Starting remote watch of plans")
+
+		var connInfo config.ConnectionInfo
+
+		if err := config.Parse(a.cfg.ConnectionInfoFile, &connInfo); err != nil {
+			logrus.Fatalf("Unable to parse connection info file %v", err)
+		}
+
+		k8splan.Watch(ctx, *applier, connInfo)
```

Comment:
Does this watcher handle its own errors?

Reply:
This is how it's used upstream and it does handle errors.